ctapipe/reco/event_classifier.py

```diff
@@ -104,4 +103,127 @@ def compute_Qfactor(self, proba, labels: int, nbins):
         Q = eps_g / np.sqrt(eps_h)
         Q[:] = Q[::-1].copy()
 
-        return Q, bins[1:]  # , eps_g[::-1], eps_h[::-1]
\ No newline at end of file
+        return Q, bins[1:]  # , eps_g[::-1], eps_h[::-1]
+
+    def _hyperBinning(self, x, featsToGroupBy: list):
+        """
+        This function is for hyper binning with pandas.
+        It is intended to be used here in order to level number of events before training the classifier;
+        for more general purposes, it is the Histogram in utils/fitshistogram.py to be used.
+
+        This outputs the input array grouped in as many bins as present in the featsToGroupBy list.
+        This is a list of dictionaries, each dictionary is related to a feature (array column) to be binned.
+
+        Beware of bins generation: since np.linspace is used, if you want log bins
+        you have to make log of the input array column and pass log(max and min)!
+
+        Parameters
+        ----------
+        x: input array to be binned
+
+        featsToGroupBy: list(dictionary)
+              list of dictionaries (see Examples for dict keys)
+
+        Output
+        ----------
+        A pandas.core.groupby.DataFrameGroupBy object
+
+
+        Examples
+        --------
+
+        >>> featsToGroupBy = [{'feat': 'feat0', 'maxf': max(x[:, 0]), 'minf': min(x[:, 0]), 'col': 0, 'nbins': 10},
+        ... {'feat': 'feat1', 'maxf': max(x[:, 1]), 'minf': min(x[:, 1]), 'col': 1, 'nbins': 5}]
+
+        where
+
+        'feat' is feature name (optional);
+        'maxf' and 'minf' are the range where to bin in
+        'col' is feature column number in the input array
+        'nbins' is the number of requested bins
+        """
+        dfx = pd.DataFrame(x)
+        binning_list = []
+
+        for i in range(len(featsToGroupBy)):
+            feat_dict = featsToGroupBy[i]
+            bins = np.linspace(feat_dict['minf'], feat_dict['maxf'], feat_dict['nbins']+1)
+            binning_list.append(pd.cut(dfx[feat_dict['col']], bins))
+
+        groups = dfx.groupby(binning_list)
+
+        return groups
+
+    def level_populations(self, group_gamma, group_hadron, gamma_evts, hadron_evts):
+        """Equalize number of entries in each bin.
```

Comment:
Keep in mind we may also be looking for electrons or other types events (Iron, etc) for some use cases, so perhaps naming this "signal" and "background" rather than "gamma" and "hadron" is better?  I think at least we can so far assume we have only 2 labels.